llama-index-integrations/vector_stores/llama-index-vector-store-firestore/llama_index/vector_store/firestore/base.py

```diff
@@ -0,0 +1,218 @@
+            firestore_filters.append(FieldFilter(f.key, "==", f.value))
+
+    if len(firestore_filters) == 1:
+        return firestore_filters[0]
+    if standard_filters.condition == FilterCondition.AND:
+        return And(filters=firestore_filters)
+    if standard_filters.condition == FilterCondition.OR:
+        return Or(filters=firestore_filters)
+
+    return firestore_filters
+
+
+class FirestoreVectorStore(BasePydanticVectorStore):
+    """Firestore Vector Store."""
+
+    stores_text: bool = True
+    flat_metadata: bool = True
+
+    collection_name: str
+    batch_size: Optional[int] = DEFAULT_BATCH_SIZE
+    embedding_key: Optional[str] = "embedding"
+    text_key: Optional[str] = "text"
+    metadata_key: Optional[str] = "metadata"
+    distance_strategy: Optional[DistanceMeasure] = DistanceMeasure.COSINE
+
+    _client: Client
+
+    def __init__(
+        self,
+        client: Optional[Client] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        super().__init__(**kwargs)
+        object.__setattr__(self, "_client", client or Client())
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "FirestoreVectorStore"
+
+    @property
+    def client(self) -> Any:
+        return self._client
+
+    def add(
+        self,
+        nodes: List[BaseNode],
+    ) -> List[str]:
+        """Add nodes to vector store."""
+        ids = []
+        entries = []
+        for node in nodes:
+            node_id = node.node_id
+            metadata = node_to_metadata_dict(
+                node,
+                remove_text=not self.stores_text,
+                flat_metadata=self.flat_metadata,
+            )
+            entry = {
+                self.embedding_key: node.get_embedding(),
+                self.metadata_key: metadata,
+            }
+            ids.append(node_id)
+            entries.append(entry)
+        self._upsert_batch(entries, ids)
+        return ids
+
+    def delete(self, ref_doc_id: str, **delete_kwargs: Any) -> None:
+        """
+        Delete nodes using with ref_doc_id."""
+        self._client.collection(self.collection_name).document(ref_doc_id).delete()
+
+    def query(self, query: VectorStoreQuery, **kwargs: Any) -> VectorStoreQueryResult:
+        """Query vector store."""
+        if query.query_embedding is None:
+            raise ValueError("Query embedding is required.")
+
+        k = kwargs.get("k") or DEFAULT_TOP_K
+        filters = _to_firestore_filter(query.filters) if query.filters else None
+        results = self._similarity_search(
+            query.query_embedding, k, filters=filters, **kwargs
+        )
+
+        top_k_ids = []
+        top_k_nodes = []
+
+        for result in results:
+            result_dict = result.to_dict() or {}
+            if self.metadata_key:
+                metadata = result_dict.get(self.metadata_key)
+
+            node = metadata_dict_to_node(metadata)
```

Comment:
if self.metadata_key doesn't exists , `metadata` variable doesn't exists.
I don't see it defined/initialized before this statement.  
Can we add a unit test for this as well?

Reply:
If metadata is empty it will throw a ValueError, [the key is required and has a default value](https://github.com/pr-Mais/llama_index/blob/4bfc712fc323e3acbad216b80334e5c71b3386b9/llama-index-integrations/vector_stores/llama-index-vector-store-firestore/llama_index/vector_store/firestore/base.py#L100).

https://github.com/run-llama/llama_index/blob/c24a2961aba0c5fd950787f0a8ef34cd0b366685/llama-index-core/llama_index/core/vector_stores/utils.py#L65-L70